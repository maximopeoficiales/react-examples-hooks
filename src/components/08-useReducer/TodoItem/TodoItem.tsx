import { Todo } from "../todoReducer";

interface MyProps {
  todo: Todo;
  i: number;
  toggleTodo(todo: Todo): void;
  deleteTodo(todo: Todo): void;
}
const defaultProps = {
  toggleTodo: () => {},
  deleteTodo: () => {},
};
const TodoItem = (props: MyProps) => {
  props = { ...defaultProps, ...props };
  const { todo,  toggleTodo, deleteTodo, i } = props;
  return (
    <p
      onClick={() => {
        toggleTodo(todo);
      }}
      key={todo.id}
      className={`item d-flex justify-content-between align-items-center ${
        todo.done && " complete"
      }`}
    >
      <b>
        {i + 1}. {todo.desc}
      </b>
      <button
        className="btn btn-sm btn-danger"
        onClick={() => {
          deleteTodo(todo);
        }}
      >
        Delete
      </button>
    </p>
  );
};

export default TodoItem;
