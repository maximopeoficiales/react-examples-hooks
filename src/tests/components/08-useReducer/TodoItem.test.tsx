import { shallow } from "enzyme";
import React from "react";
import TodoItem from "../../../components/08-useReducer/TodoItem/TodoItem";
import { newTodo } from "../../fixtures/demoTodos";

describe("Pruebas en TodoItem.tsx", () => {
  const index = 1;
  const deleteTodo = jest.fn();
  const toggleTodo = jest.fn();
  const wrapper = shallow(
    <TodoItem
      deleteTodo={deleteTodo}
      toggleTodo={toggleTodo}
      i={index}
      todo={newTodo}
    />
  );

  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de llamar a la funcion handleDelete", () => {
    // simulo interaccion con el boton delete
    wrapper.find("button").simulate("click");
    expect(deleteTodo).toHaveBeenCalled();
    expect(deleteTodo).toHaveBeenCalledTimes(1);
    expect(deleteTodo).toHaveBeenCalledWith(newTodo);
  });
  test("debe de llamar a la funcion handleToggle", () => {
    // simulo interaccion con el parrafo
    wrapper.find("p").simulate("click");
    expect(toggleTodo).toHaveBeenCalled();
    expect(toggleTodo).toHaveBeenCalledTimes(1);
    expect(toggleTodo).toHaveBeenCalledWith(newTodo);
  });
  test("debe mostrar el texto correctamente", () => {
    let texto = wrapper.find("b");
    // console.log(texto.html());
    expect(texto.text()).toEqual(`${index + 1}. ${newTodo.desc}`);
  });

  test("debe mostrar la class complete cuando todo.done = true", () => {
    expect(wrapper.find("p").hasClass("complete")).toBeTruthy();
    console.log(wrapper.html());
  });
});
