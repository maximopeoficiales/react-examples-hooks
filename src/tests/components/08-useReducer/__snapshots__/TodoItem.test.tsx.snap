// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoItem.tsx debe de mostrarse correctamente 1`] = `
<p
  className="item d-flex justify-content-between align-items-center  complete"
  key="2"
  onClick={[Function]}
>
  <b>
    2
    . 
    alguna descripcion 2
  </b>
  <button
    className="btn btn-sm btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</p>
`;
